File: Problem-3.js
// Time Complexity : O(n)
// Space Complexity : O(2n) = O(n)
// Did this code successfully run on Leetcode : Yes (Problem: 290)
// Any problem you faced while coding this : None


// Your code here along with comments explaining your approach

/**
 * @param {string} pattern
 * @param {string} s
 * @return {boolean}
 */
 var wordPattern = function(pattern, s) {
    // Convert into array
        const secondString = s.split(" ")
        
        // If both length are not same
        if (pattern == null || s == null || pattern.length != secondString.length) return false;
    
        const a = {}, b = {};
        for (let i = 0; i < pattern.length; i++) {
            // If value does not exist in object or undefined
            if (a[pattern[i]] !== b[secondString[i]]) {
                return false;
            }
            a[pattern[i]] = i + 1;
            b[secondString[i]] = i + 1;
        }
        return true
    };

File: problem_3.cpp
/ Time Complexity : O(n)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this :

class Solution {
    vector<string> splitString(string s) {
        vector<string> vec;
        string st = "";
        
        for (int i = 0; i < s.length(); i++) {
            if (s[i] == ' ') {
                vec.push_back(st);
                st = "";
            } else {
                st += s[i];
            }
        }
        vec.push_back(st);
        return vec;
    }
    
public:
    bool wordPattern(string pattern, string s) {
            vector<string> s_word_vec = splitString(s);
            vector<char> pattern_vec(pattern.begin(), pattern.end());
            
            if(s_word_vec.size() != pattern_vec.size()) return false;
        
            unordered_map<char, string> umap;
            set<string> se;
         
            for(int i = 0; i < s_word_vec.size(); i++) {
                //umap contains mapping
                if (umap.find(pattern_vec[i]) != umap.end()) {
                    if (umap[pattern_vec[i]] != s_word_vec[i]) {
                        return false;
                    }
                } else {
                    if (se.find(s_word_vec[i]) != se.end()) return false;
                }
                
                umap.insert(make_pair(pattern_vec[i], s_word_vec[i]));
                se.insert(s_word_vec[i]);
            }
        return true;
    }
};
